include/netlink/list.h

```diff
@@ -46,6 +46,35 @@ static inline void nl_list_add_head(struct nl_list_head *obj,
 	__nl_list_add(obj, head, head->next);
 }
 
+static inline void nl_list_insert_before(struct nl_list_head *obj,
+					 struct nl_list_head *ref)
+{
+	__nl_list_add(obj, ref->prev, ref);
+}
+
+static inline void nl_list_insert_after(struct nl_list_head *obj,
+					struct nl_list_head *ref)
+{
+	__nl_list_add(obj, ref, ref->next);
+}
+
+static inline void nl_list_insert_list_after(struct nl_list_head *head,
+					     struct nl_list_head *ref)
+{
+	ref->next->prev = head->prev;
+	head->prev->next = ref->next;
+	ref->next = head->next;
+	head->next->prev = ref;
+	head->next = head;
+	head->prev = head;
+}
+
+static inline void nl_list_join(struct nl_list_head *head_1,
+				struct nl_list_head *head_2)
+{
+	nl_list_insert_list_after(head_2, head_1->prev);
```

Comment:
if the code is copied from linux kernel, then we must take care to not use invalid licenses (libnl is LGPL2.1-only).

Granted, those functions seem so minimal, I don't think you can implement them any other way.

If you want to get inspired, then compare to https://github.com/c-util/c-list/blob/main/src/c-list.h which has a compatible license. The downside is, we really want that the API looks similar to what is in kernel, so that somebody familiar with kernel API understand it.

Use your best judgment :)

Reply:
I would merge this patch early, even if no users are merged yet. I think it's generally useful API beyond this PR.

Reply:
I agree with all your points. I myself found it hard to understand how the list module worked since it had no documentation. It also didn't help that I'd never seen that type of implementation of a linked list before. I'll make a separate PR adds the extra methods and documents the module. I can add some unit tests as well if you could give me some pointers on how to do that.

> Does this code come from linux kernel headers

No it doesn't. I hadn't looked at the code in the Linux kernel for their list implementation until now. I wrote the code based on a method I worked out on paper. The closest thing the kernel has seems to be `list_splice_tail_init` in `include/linux/list.h`. You seem to be able to use that function to do the same thing. It just has the parameters swapped around I think.

Reply:
> I've checked and the library provides a means to do everything we can currently do with our implementation, except it it doesn't have reverse iteration macros. Those would be easy enough to add, but that might mean the code would have to copied over instead of using a subtree (or another file includes that file to add to it).

I think the point would be that we submit such features to c-list "upstream", and just do a `git subtree pull`. That requires that upstream is susceptible to such patches. I think upstream is, otherwise there would be a problem with pulling it.

Thereby others could potentially benefit to that contribution, not just some obscure part inside libnl3.

> Of course it will still take some work to convert the existing code to use the new interface.

I don't think we need to rework existing code (unless somebody wants to send patches). It could be used only for new code.

Reply:
how about https://github.com/thom311/libnl/pull/412 ?

Reply:
Looks good. I'll also make a PR that adds reverse iterators to that library. I've added the code; I've just got to adapt some of their tests. I'll probably submit that tomorrow.

Reply:
Here's the PR I made, for reference: https://github.com/c-util/c-list/pull/14

Reply:
lgtm.

In the meantime, for your WIP branch here, I think you can do:

```
git subtree pull --prefix third_party/c-list/ git@github.com:c-util/c-list.git refs/pull/14/head --squash
```